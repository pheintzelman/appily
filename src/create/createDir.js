const fsModule = require('fs');
const fs = fsModule.promises;
const { DefaultAppName } = require('../constants');
const { logger } = require('../logger');

async function createDir(config) {
  const name = config.name ?? DefaultAppName;
  await fs.mkdir(name);
  logger.debug(`Dir created: ${name}`);
}

module.exports = { createDir };
