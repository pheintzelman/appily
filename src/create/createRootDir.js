import fsModule from 'fs';
const fs = fsModule.promises;
import { DefaultAppName } from '../constants.js';
import { logger } from '../logger.js';
import { dirExists } from '../lib/file.js';

// If the dir already exists append the next number
async function getDirName(name, count = 0) {
  const dir = count === 0 ? name : `${name}${count}`;

  if ((await dirExists(dir)) && count <= 100) {
    return getDirName(name, count + 1);
  }

  return dir;
}

export async function createRootDir(config) {
  const name = config.name ?? DefaultAppName;
  const dir = await getDirName(name);

  await fs.mkdir(dir);
  logger.debug(`Dir created: ${dir}`);
  return dir;
}
