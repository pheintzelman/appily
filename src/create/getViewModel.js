import util from 'util';
import { getVariations } from '../lib/case.js';
import { Type } from '../constants/constants.js';
import { getDefaultValueForType } from './viewModel/getDefaultValueForType.js';
import { mapTypeToComponent } from './viewModel/mapTypeToComponent.js';
import { getFormImports } from './viewModel/getFormImports.js';

function getDefaultState(properties) {
  return properties.reduce((acc, { propertyNameCamel, type }) => {
    return { ...acc, [propertyNameCamel]: getDefaultValueForType(type) };
  }, {});
}

function typeFlags(type) {
  return { isString: type === Type.String, isBoolean: type === Type.Boolean };
}

function preprocessProperty([propertyName, property]) {
  return {
    ...getVariations('propertyName', propertyName),
    type: property,
    ...typeFlags(property),
    component: mapTypeToComponent(property)
  };
}

function preprocessProperties(properties) {
  if (!properties) {
    return [];
  }

  return Object.entries(properties).map(preprocessProperty);
}

function preprocessModel([modelName, model]) {
  const properties = preprocessProperties(model);

  return {
    properties,
    defaultState: util.inspect(getDefaultState(properties)),
    ...getVariations('modelName', modelName),
    formImports: getFormImports(properties)
  };
}

function preprocessModels({ config, manifesto, dir }) {
  if (!config.models) {
    return [];
  }

  return Object.entries(config.models).map(preprocessModel);
}

export function getViewModel({ config, manifesto, dir }) {
  return {
    ...getVariations('appName', config.name),
    dir,
    templateName: manifesto.name,
    templateVersion: manifesto.version,
    models: preprocessModels({ config, manifesto, dir }),
    configString: JSON.stringify(config, null, 2)
  };
}
