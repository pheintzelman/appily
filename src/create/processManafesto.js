import { kvp } from '../lib/kvp.js';

export function getSelectedDirectives(manifesto, selectedOptions) {
  const { directives } = manifesto;
  return directives
    .map(kvp)
    .filter((directive) => selectedOptions.includes(directive.key))
    .map((directive) => directive.value)
    .map(kvp);
}

export function getSelectedOptions(manifesto, config) {
  const { options } = manifesto;

  return Object.entries(options)
    .filter(([key, value]) => value.length >= 1)
    .map(([key, value]) => {
      if (value.includes(config[key])) {
        return `${key}:${config[key]}`;
      }

      //include the first option by default
      return `${key}:${value[0]}`;
    });
}

export function getDirectives(manifesto, config) {
  const selectedOptions = getSelectedOptions(manifesto, config);
  return getSelectedDirectives(manifesto, selectedOptions);
}
