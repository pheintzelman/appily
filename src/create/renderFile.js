import fsModule from 'fs';
const fs = fsModule.promises;
import mustache from 'mustache';
import { logger } from '../logger.js';
import { TemplateExt } from '../constants.js';

function renameDest(dest) {
  return dest.replace(TemplateExt, '');
}

export async function renderFile({ src, dest: destIn, viewModel }) {
  const mustacheFile = await fs.readFile(src, 'utf8');
  const file = mustache.render(mustacheFile, viewModel);
  const dest = renameDest(destIn);

  await fs.writeFile(dest, file, 'utf8');
  logger.debug(`File rendered: ${dest}`);
}
