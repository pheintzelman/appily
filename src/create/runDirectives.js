import path from 'path';
import { fileURLToPath } from 'url';
import { Directive, ManifestoFileName } from '../constants.js';
import { logger } from '../logger.js';
import { kvp } from '../lib/kvp.js';
import { readJsonFile } from '../lib/file.js';
import { copyDirective } from './directives/copyDirective.js';
import { loopDirective } from './directives/loopDirective.js';
import { viewModelDirective } from './directives/viewModelDirective.js';
import { getViewModel } from './getViewModel.js';
import { access } from 'fs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

function getDirectives(manifesto, selectedOptions) {
  const { directives } = manifesto;
  return directives
    .map(kvp)
    .filter((directive) => selectedOptions.includes(directive.key))
    .map((directive) => directive.value);
}

function getSelectedOptions(manifesto, config) {
  const { options } = manifesto;

  return Object.entries(options)
    .filter(([key, value]) => value.length >= 1)
    .map(([key, value]) => {
      if (value.includes(config[key])) {
        return `${key}:${config[key]}`;
      }

      //include the first option by default
      if (value.length >= 1) {
        return `${key}:${value[0]}`;
      }
    });
}

async function readManifesto(templateDir) {
  const manifestoPath = path.join(templateDir, ManifestoFileName);
  return await readJsonFile(manifestoPath);
}

async function runFileDirectives({ directives, templateDir, dir, viewModel }) {
  for (const directive of directives) {
    if (directive.key === Directive.Copy) {
      const src = path.join(templateDir, directive.value.src);
      const dest = path.join(dir, directive.value.dest);
      await copyDirective({ src, dest, viewModel });
    }

    if (directive.key === Directive.Loop) {
      const src = path.join(templateDir, directive.value.src);
      const dest = path.join(dir, directive.value.dest);
      const { property } = directive.value;

      await loopDirective({ property, src, dest, viewModel });
    }
  }
}

function runViewModelDirectives({ directives, initialViewModel }) {
  return directives.reduce((viewModel, directive) => {
    if (directive.key === Directive.ViewModel) {
      const property = directive.value;
      return viewModelDirective({ property, viewModel });
    }

    return viewModel;
  }, initialViewModel);
}

export async function runDirectives(config, dir) {
  const { template } = config;
  const templateDir = path.resolve(__dirname, `../templates/${template}`);
  const manifesto = await readManifesto(templateDir);
  const selectedOptions = getSelectedOptions(manifesto, config);
  const directives = getDirectives(manifesto, selectedOptions).map(kvp);
  const initialViewModel = getViewModel({ config, manifesto, dir });
  const viewModel = runViewModelDirectives({ directives, initialViewModel });
  logger.trace({ viewModel });

  await runFileDirectives({ directives, templateDir, dir, viewModel });
}
