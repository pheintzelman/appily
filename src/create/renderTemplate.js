import fsModule from 'fs';
const fs = fsModule.promises;
import path from 'path';
import { fileURLToPath } from 'url';
import { logger } from '../logger.js';
import { dirExists } from '../lib/file.js';
import { renderFile } from './renderFile.js';
import { TemplateExt } from '../constants.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

async function copyFile({ src, dest, viewModel }) {
  const isTemplateFile = src.includes(TemplateExt);
  if (isTemplateFile) {
    return await renderFile({ src, dest, viewModel });
  }

  await fs.copyFile(src, dest);
  logger.debug(`File copied: ${dest}`);
}

async function copyTemplate({ src, dest, viewModel }) {
  logger.trace({ msg: 'copyDir', src, dest });

  const entries = await fs.readdir(src, { withFileTypes: true });

  if (!(await dirExists(dest))) {
    await fs.mkdir(dest);
  }

  for (const entry of entries) {
    const srcPath = path.join(src, entry.name);
    const destPath = path.join(dest, entry.name);
    if (entry.isDirectory()) {
      await copyDir(srcPath, destPath);
    } else {
      await copyFile({ src: srcPath, dest: destPath, viewModel });
    }
  }
}

export async function renderTemplate(config, dest) {
  const { template } = config;
  const templateDir = path.resolve(__dirname, `../templates/${template}`);
  const src = path.join(templateDir, 'app');
  const viewModel = { appName: config.name };

  await copyTemplate({ src, dest, viewModel });
}
