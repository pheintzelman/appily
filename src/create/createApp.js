import path from 'path';
import { ManifestoFileName } from '../constants/constants.js';
import { createRootDir } from './file/createRootDir.js';
import { runDirectives } from './runDirectives.js';
import { validateConfig } from './config/validateConfig.js';
import { readJsonFile } from '../lib/file.js';
import { getTemplateDir } from './file/getTemplateDir.js';
import { logger } from '../logger.js';
import { normalizeConfig } from './config/normalizeConfig.js';

async function readManifesto(templateDir) {
  const manifestoPath = path.join(templateDir, ManifestoFileName);
  return await readJsonFile(manifestoPath);
}

export async function createApp(config) {
  const templateDir = getTemplateDir(config);
  const manifesto = await readManifesto(templateDir);

  const normalizedConfig = normalizeConfig(config);
  logger.trace({ normalizedConfig });
  await validateConfig(normalizedConfig, manifesto);
  const dir = await createRootDir(normalizedConfig);
  await runDirectives({
    config: normalizedConfig,
    dir,
    manifesto,
    templateDir
  });
  logger.info(`${normalizedConfig.name} created, in dir ${dir}`);
}
