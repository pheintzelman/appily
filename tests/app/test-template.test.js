import { createApp } from '../../src/create/createApp';
import {
  removeDir,
  getSanpshotTestCasesSync
} from './helpers/snapshot-helpers.js';
import { captureLog, sortLog } from '../helpers/log.js';

const config = {
  dir: 'tests/app',
  name: 'Test App',
  template: 'test'
};

const log = captureLog();
await removeDir('tests/app/test-app');
await createApp(config);

describe('Test template', () => {
  const testCases = getSanpshotTestCasesSync('tests/app/test-app');

  test.each(testCases)(
    '$dir should match snapshot',
    ({ name, dir, snapshot }) => {
      expect(snapshot).toMatchSnapshot();
    }
  );

  test('should match logs', () => {
    expect(sortLog(log)).toMatchSnapshot();
  });
});
