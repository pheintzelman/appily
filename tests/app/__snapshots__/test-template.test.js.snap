// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test template should match logs 1`] = `
Array [
  Object {
    "TRACE": Object {
      "viewModel": Object {
        "appName": "Test App",
        "appNameCamel": "testApp",
        "appNamePascal": "TestApp",
        "appNameSnake": "test-app",
        "configString": "{
  \\"dir\\": \\"tests/app\\",
  \\"name\\": \\"Test App\\",
  \\"template\\": \\"test\\",
  \\"models\\": {}
}",
        "dir": "tests/app/test-app",
        "models": Array [],
        "templateName": "test",
        "templateVersion": "0.0.1",
      },
    },
  },
  Object {
    "TRACE": Object {
      "normalizedConfig": Object {
        "dir": "tests/app",
        "models": Object {},
        "name": "Test App",
        "template": "test",
      },
    },
  },
  Object {
    "TRACE": Object {
      "dest": "tests/app/test-app/UI",
      "msg": "mkdir",
    },
  },
  Object {
    "TRACE": Object {
      "dest": "tests/app/test-app/UI",
      "msg": "copyDir",
      "src": "/Users/paulheintzelman/Documents/projects/create-app/src/templates/test/app/UI",
    },
  },
  Object {
    "TRACE": Object {
      "dest": "tests/app/test-app",
      "msg": "copyDir",
      "src": "/Users/paulheintzelman/Documents/projects/create-app/src/templates/test/app",
    },
  },
  Object {
    "INFO": "Test App created, in dir tests/app/test-app",
  },
  Object {
    "DEBUG": "File rendered: tests/app/test-app/UI/index.html",
  },
  Object {
    "DEBUG": "File copied: tests/app/test-app/helloWorld.js",
  },
  Object {
    "DEBUG": "Dir created: tests/app/test-app",
  },
]
`;

exports[`Test template tests/app/test-app/UI/index.html should match snapshot 1`] = `
"<html>
  <body>
    <h1>Welcome to Test App!</h1>
  </body>
</html>
"
`;

exports[`Test template tests/app/test-app/helloWorld.js should match snapshot 1`] = `
"function helloWorld() {
  console.log('Hello World');
}
"
`;
